### PR TITLE
Improved event reports

### PR DESCRIPTION
**Pull request by jusrus01**
Added pagination, sorting and filtering by date to event report participant visited events.

### CODE REVIEW

File: src/api/Shrooms.Premium/Domain/Services/Events/List/EventListingService.cs [hunk 1]
```diff
@@ -188,7 +210,12 @@ public async Task<IPagedList<EventParticipantReportDto>> GetReportParticipantsAs
                             StartDate = visited.StartDate,
                             EndDate = visited.EndDate
                         })
-                        .ToList()
+                        .ToList(),
+                    TotalVisitedEventCount = _eventsDbSet
+                        .Count(e => e.EventParticipants.Any(participant => participant.ApplicationUserId == p.ApplicationUser.Id) &&
+                                    e.EndDate < DateTime.UtcNow &&
+                                    e.OrganizationId == userOrg.OrganizationId &&
+                                    (eventTypesLength == 0 || reportArgsDto.EventTypeIds.Contains(e.EventType.Id)))
```
Comment (jusrus01) [inline src/api/Shrooms.Premium/Domain/Services/Events/List/EventListingService.cs:218]: Tried to export this LINQ to Expression<Func<Event, bool>> (in this method we have the same expression twice) function, however I kept getting this error.
Screenshot 2022-07-14 104859

Comment (SauliusStankevicius) [reply src/api/Shrooms.Premium/Domain/Services/Events/List/EventListingService.cs:218]: Which EF version does Simoona use? it seems there is a bug with making queries like that for navigation properties in EF 6, and also what happens if you only move the bool check inside the Any() to a separate method?

File: src/api/Shrooms.Premium/Presentation/WebViewModels/ValidationAttributes/DateTimeLessThanPresentDateAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+using System.ComponentModel.DataAnnotations;
+
+namespace Shrooms.Premium.Presentation.WebViewModels.ValidationAttributes
+{
+    [AttributeUsage(AttributeTargets.Property)]
+    public class DateTimeLessThanPresentDateAttribute : ValidationAttribute
+    {
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            if (value == null || ((DateTime)value) <= DateTime.UtcNow)
```
Comment (SauliusStankevicius) [inline src/api/Shrooms.Premium/Presentation/WebViewModels/ValidationAttributes/DateTimeLessThanPresentDateAttribute.cs]: Does UI send dates in UTC? and maybe better to validate if value is of type DateTime first, as this one can throw exception if attribute was set on some other property

File: src/api/Shrooms.Premium/Presentation/WebViewModels/Events/EventParticipantVisitedEventsListingArgsViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+        [Range(1, int.MaxValue)]
+        public int PageSize { get; set; } = EventsConstants.EventsDefaultPageSize;
+        
+        public string SortByProperties { get; set; }
+
+        public IEnumerable<int> EventTypeIds { get; set; }
+
+        [Required]
+        [DateTimeLessThanDateTime("EndDate")]
```
Comment (SauliusStankevicius) [inline src/api/Shrooms.Premium/Presentation/WebViewModels/Events/EventParticipantVisitedEventsListingArgsViewModel.cs]: could `nameof` be used in here?

File: src/api/Shrooms.Premium/Presentation/WebViewModels/ValidationAttributes/DateTimeLessThanPresentDateAttribute.cs (hunk 2 above)
Comment (jusrus01) [reply src/api/Shrooms.Premium/Presentation/WebViewModels/ValidationAttributes/DateTimeLessThanPresentDateAttribute.cs]: It does send UTC. At least everything related with events and profile editing. Also, is that check necessary when the attribute name clearly states that it has to be used with DateTime?

Comment (SauliusStankevicius) [reply src/api/Shrooms.Premium/Presentation/WebViewModels/ValidationAttributes/DateTimeLessThanPresentDateAttribute.cs]: the problem is that nobody is preventing it to be used on different property type

File: src/api/Shrooms.Premium/Domain/Services/Events/List/EventListingService.cs (hunk 1 above)
Comment (jusrus01) [reply src/api/Shrooms.Premium/Domain/Services/Events/List/EventListingService.cs:218]: Version is 6.1.3. 

Exporting Any() I get:
image

**Review by SauliusStankevicius — APPROVED** [review]

**Pull request merged**